src/functional/geometry/convex_hull.zig

```diff
@@ -0,0 +1,167 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Implementation of the Graham's scan which can maintain a convex hull from the paper
+//! "Graham, R.L. An Efficient Algorithm for Determining the Convex Hull of a Finite Planar Set.
+//! Information Processing Letters. 1 (4): 132–133. 1972.
+//! https://doi.org/10.1016/0020-0190(72)90045-2.
+//! Convex Hulls are used by the Slide Filter and Optimal PLA algorithms implemented in
+//! `src/functional/swing_slide_filter.zig` and `src/functional/optimal_pla.zig`.
+
+const std = @import("std");
+const mem = std.mem;
+const ArrayList = std.ArrayList;
+const testing = std.testing;
+
+const tersets = @import("../../tersets.zig");
+const Point = tersets.Point;
+
+/// Enum for the angle's `Turn` of three consecutive points A, B, and C. Essentially, it describes
+/// whether the path from A to B to C makes a `left` turn, a `right` turn, or continues in a
+/// straight line also called collinear.
+const Turn = enum(i8) { right, left, collinear };
+
+/// Convex Hull formed by an upper and lower hull.
+pub const ConvexHull = struct {
+    lower_hull: ArrayList(Point),
+    upper_hull: ArrayList(Point),
+
+    // Initialize the container with a given `allocator`.
+    pub fn init(allocator: mem.Allocator) !ConvexHull {
+        return ConvexHull{
+            .lower_hull = ArrayList(Point).init(allocator),
+            .upper_hull = ArrayList(Point).init(allocator),
+        };
+    }
+
+    // Deinitialize the container and free the allocated memory.
+    pub fn deinit(self: *ConvexHull) void {
+        self.lower_hull.deinit();
+        self.upper_hull.deinit();
+    }
+
+    /// Add new `point` to the convex hull.
+    pub fn addPoint(self: *ConvexHull, point: Point) !void {
+        try addPointToHull(&self.upper_hull, Turn.right, point);
+        try addPointToHull(&self.lower_hull, Turn.left, point);
+    }
+};
+
+/// Auxiliary function to add a new `point` to a given `hull` of the convex hull. The function uses
+/// the given `turn` to correctly add the new point.
+fn addPointToHull(hull: *ArrayList(Point), turn: Turn, point: Point) !void {
+    if (hull.items.len < 2) {
+        // The first two points can be add directly.
+        try hull.append(point);
+    } else {
+        var top: usize = hull.items.len - 1;
+        // Remove the last point as long as the `turn` is not the provided.
+        while ((top > 0) and (computeTurn(
+            hull.items[top - 1],
+            hull.items[top],
+            point,
+        ) != turn)) : (top -= 1) {
+            _ = hull.pop();
+        }
+        try hull.append(point);
+    }
+}
+
+/// Compute turn created by the path from `first_point` to `middle_point` to `last_point`.
+fn computeTurn(first_point: Point, middle_point: Point, last_point: Point) Turn {
```

Comment:
Maybe add another sentence to the comment making it clear why `computeTurn()` is not part of the `ConvexHull` despite being a helper functions for methods in that type. Seems it should be possible to implement it as a private method that does not take `self` as the first argument like `ArrayList.init()` as shown in the Zig documentation [here](https://ziglang.org/documentation/master/#struct).

Reply:
Well, I just tried declaring the function inside the structure without `self` as the first argument and everything works well. So I fixed this comment by bringing the function inside and slightly changing the documentation.